### PR TITLE
spirv-fuzz: fuzzer pass to adjust memory access operands

### LINKED ISSUE #2940: spirv-fuzz: add memory access operand
It is always harmless, semantics-wise, to apply the "Volatile" or "Nontemporal" memory access operands to instructions that manipulate memory (e.g. OpLoad).

"Volatile" should not be removed - it could be needed.

"Nontemporal" could be removed - it's just a hint.

The fuzzer should play with these.

### PR DESCRIPTION
**Pull request by afd**
A new pass that gives spirv-fuzz the ability to adjust the memory
operand masks associated with memory access instructions (such as
OpLoad and OpCopy Memory).

Fixes #2940.

### CODE REVIEW

File: source/fuzz/fuzzer_pass_adjust_memory_operands_masks.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+          bool toggle_nontemporal =
+              !add_volatile || GetFuzzerContext()->ChooseEven();
+
+          // These bitwise operations use '|' to add Volatile if desired, and
+          // '^' to toggle Nontemporal if desired.
+          uint32_t new_mask =
+              (existing_mask | (add_volatile ? SpvMemoryAccessVolatileMask
+                                             : SpvMemoryAccessMaskNone)) ^
+              (toggle_nontemporal ? SpvMemoryAccessNontemporalMask
+                                  : SpvMemoryAccessMaskNone);
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_adjust_memory_operands_masks.cpp]: I suppose you could extract this into a public static method and then add a test for it?

File: source/fuzz/instruction_descriptor.cpp [hunk 2]
```diff
@@ -66,5 +66,35 @@ protobufs::InstructionDescriptor MakeInstructionDescriptor(
...
+       backwards_iterator != block.begin(); --backwards_iterator) {
+    if (backwards_iterator->HasResultId()) {
+      // As soon as we find an instruction with a result id, we can return a
+      // descriptor for |inst_it|.
+      return MakeInstructionDescriptor(backwards_iterator->result_id(), opcode,
+                                       skip_count);
+    }
+    if (backwards_iterator != inst_it &&
+        backwards_iterator->opcode() == opcode) {
+      // We are skipping over an with the same opcode as |inst_it|; we increase
```
Comment (paulthomson) [inline source/fuzz/instruction_descriptor.cpp]: typo: an instruction

File: source/fuzz/protobufs/spvtoolsfuzz.proto [hunk 3]
```diff
@@ -464,6 +465,25 @@ message TransformationSetLoopControl {
...
+message TransformationSetMemoryOperandsMask {
+
+  // A transformation that sets the memory operands mask of a memory access
+  // instruction.
+
+  // A descriptor for a memory access instruction, e.g. an OpLoad
+  InstructionDescriptor memory_access_instruction = 1;
+
+  // A mask of memory operands to be applied to the instruction.  It must be the
+  // same as the original mask, except that Volatile can be removed, and
```
Comment (paulthomson) [inline source/fuzz/protobufs/spvtoolsfuzz.proto]: Volatile can be added

File: source/fuzz/transformation_set_memory_operands_mask.cpp [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  }
+  if (!IsMemoryAccess(*instruction)) {
+    return false;
+  }
+
+  if (message_.memory_operands_mask_index() != 0) {
+    assert(message_.memory_operands_mask_index() == 1);
+    assert(instruction->opcode() == SpvOpCopyMemory ||
+           instruction->opcode() == SpvOpCopyMemorySized);
+    assert(MultipleMemoryOperandMasksAreSupported(context));
```
Comment (paulthomson) [inline source/fuzz/transformation_set_memory_operands_mask.cpp]: I wonder if these should be more than just debug assertions. Do we assume we won't accidentally find a different instruction?

File: source/fuzz/transformation_set_memory_operands_mask.cpp [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  auto original_mask_in_operand_index = GetInOperandIndexForMask(
+      *instruction, message_.memory_operands_mask_index());
+  assert(original_mask_in_operand_index != 0 &&
+         "The given mask index is not valid.");
```
Comment (paulthomson) [inline source/fuzz/transformation_set_memory_operands_mask.cpp]: Is this right?

File: source/fuzz/instruction_descriptor.cpp (hunk 2 above)
Comment (afd) [reply source/fuzz/instruction_descriptor.cpp]: Fixed.

File: source/fuzz/protobufs/spvtoolsfuzz.proto (hunk 3 above)
Comment (afd) [reply source/fuzz/protobufs/spvtoolsfuzz.proto]: Good catch.

File: source/fuzz/transformation_set_memory_operands_mask.cpp (hunk 4 above)
Comment (afd) [reply source/fuzz/transformation_set_memory_operands_mask.cpp]: They really should never be violated.  I have moved them to the top of the method and made the accesses to 'opcode' refer to the message's instruction descriptor, and added a comment to explain.

File: source/fuzz/fuzzer_pass_adjust_memory_operands_masks.cpp (hunk 1 above)
Comment (afd) [reply source/fuzz/fuzzer_pass_adjust_memory_operands_masks.cpp]: I will leave this as is, since the code involves choosing randomly.

**Pull request merged**